### ISSUE TITLE
[RFC] A new network wire protocol

### ISSUE THREAD

**Issue by bothra90** [libra/libra#2878]
**tl;dr:** We're changing from stream-oriented communication to message-oriented communication.

## Why?

We've been operating the current network stack in Libra for almost a year with little to no issues, and acceptable performance. So why change? Here are a few reasons:

- **Performance**: The current protocol and implementation stack is based on [Yamux](https://github.com/hashicorp/yamux/blob/master/spec.md). Yamux allows us to multiplex several "streams" on a single TCP connection, enabling execution of several RPCs and message sends in parallel. It also allows for "server-initiated" streams, making it a good fit for p2p communication needed in blockchain systems. However, the protocol introduces limitations as a result of the critical design choice of using [per-stream flow control](https://github.com/hashicorp/yamux/blob/master/spec.md#message-flow). Yamux has a notion of `receive_window` which enables per-stream flow control. A receiver must issue a `receive_window_update` before the sender can send any more data on a stream once the current window has been saturated. The default assumed receive window size is `256KB`, but can be updated any time during a stream's lifetime. This, however, introduces extra latency of 1 RTT in communication whenever the sender needs to send data larger than the default window size, since it must wait for the `receive_window_update` to arrive after the 256 KB of data is sent. We believe this is unnecessary for our purpose, and that TCP's in-built flow control is sufficient. Additionally, this makes the protocol hard to debug, since sending of data blocks if the receiver has not updated the receive window.
- **Message priority**: Although Yamux provides fairness through per-stream flow control, it doesn't have a notion of stream priority. This means that all streams compete for a shared resource, i.e., the underlying connection. Having prioritized delivery is useful in cases like Libra, where the message priority is fairly deterministic (e.g. consensus messages are higher priority than mempool messages) and which can benefit from reduced latency in some parts while sacrificing latency in others. One could implement notions of priority in the Yamux libraries at the two ends of a connection, but that seems a fairly complex undertaking.
- **Safety and inter-operability with TCB networking**: The current state-of-the-art Yamux implementation in Rust exposes an `async` API and requires a `futures` executor to drive the underlying connection. Since we wish to use LibraNet for secure communication between consensus and TCB (instead of the current bespoke [TCB networking library](https://github.com/libra/libra/blob/master/secure/net/src/lib.rs)), it's desirable to avoid the need for using an executor, since executors such as `tokio` use (or depend on) a fair amount of `unsafe` code.
- **Ecosystem**: Although the Yamux protocol is now more than 5 years old, it lacks community-wide adoption. Quality implementations are only available in [Go](https://github.com/hashicorp/yamux) and [Rust](https://github.com/paritytech/yamux). Implementors of Libra in any other language would need to take on the necessary undertaking of implementing the Yamux protocol in addition to the Libra protocol. Since re-implementing Libra in as many languages as possible and increasing ecosystem diversity is something we believe is important, we think it's best to remove Yamux and instead leverage `lcs` which needs to be implemented anyhow.

## New protocol:

While complete details of the new protocol are still being flushed out, here are the basic pieces:

The core networking protocol is built around sending and receiving the fully self-contained `NetworkMessage` struct defined below:

```rust
/// Message variants that are sent on the wire.
#[derive(Clone, Debug, Deserialize, Serialize)]
pub enum NetworkMessage {
    Error(ErrorCode),
    Ping(Nonce),
    Pong(Nonce),
    RpcRequest(RpcRequest),
    RpcResponse(RpcResponse),
    DirectSendMsg(DirectSendMsg),
}

/// Enum representing various error codes that can be embedded in error messages.
#[derive(Clone, Debug, Deserialize, Serialize)]
pub enum ErrorCode {
    /// Failed to parse NetworkMessage when interpreting according to provided protocol version.
    ParsingError(MessagingProtocolVersion, Box<NetworkMessage>),
    UnsupportedProtocol(ProtocolId),
    TimedOut,
}

/// Nonces used by Ping and Pong message types.
#[derive(Clone, Debug, Deserialize, Serialize)]
pub struct Nonce(u32);

#[derive(Clone, Debug, Deserialize, Serialize)]
pub struct RpcRequest {
    // RequestId for the RPC Request.
    pub request_id: u32,
    // `protocol_id` is an enum variant of ProtocolId enum defined in the handshake
    // protocol.
    pub protocol_id: ProtocolId,
    // Request priority.
    pub priority: u8,
    // Request payload. This will be parsed by the application-level handler.
    pub raw_request: Bytes,
}

#[derive(Clone, Debug, Deserialize, Serialize)]
pub struct RpcResponse {
    // RequestId for corresponding request; copied as is from RpcRequest.
    pub request_id: u32,
    // Response priority. This will likely be same as the priority of corresponding request.
    pub priority: u8,
    // Response payload.
    pub raw_response: Bytes,
}

#[derive(Clone, Debug, Deserialize, Serialize)]
pub struct DirectSendMsg {
    // `protocol_id` is an enum variant of ProtocolId enum defined in the handshake
    // protocol.
    pub protocol_id: ProtocolId,
    // Message priority.
    pub priority: u8,
    // Message payload.
    pub raw_msg: Bytes,
}


/// Type for unique identifier associated with each network protocol
/// List of application level protocols.
#[repr(u8)]
#[derive(Clone, Copy, Debug, Hash, Eq, PartialEq, Deserialize_repr, Serialize_repr)]
pub enum ProtocolId {
    ConsensusRpc = 0,
    ConsensusDirectSend = 1,
    MempoolDirectSend = 2,
    StateSynchronizerDirectSend = 3,
    DiscoveryDirectSend = 4,
    HealthCheckerRpc = 5,
}
```

We use `lcs` for serializing the `NetworkMessage` struct on the wire and length-prefix that with a `u32`. Note that all fields are in little-endian form. Effectively, the network socket (which can be thought of as a stream of bytes) becomes a stream (and sink) of NetworkMessages. These network messages are then handled by the network module and sent to registered upstream handlers.

It is important to note that there will be *some* added complexity in the implementation of the `rpc` module which now needs to manage RPC completions instead of relying on the request and responses being embedded inside streams which are managed by the yamux library. On the other hand, we will get rid of things such as per-stream protocol negotiation (since the ProtocolId is now embedded inside the messages), maintaining long-lived streams for `DirectSend`, and the need to open a new stream per RPC.

More details, such as dependencies and protocol upgradability, will be captured in the full Libra wire spec.

cc: @phlip9, @aching, @davidiw

**Comment by davidiw** [comment]
Could you elaborate on the handshake protocol?

**Comment by davidiw** [comment]
@phlip9 any reason not to close this one?
